# Model Algorithm 

#### Linear Regressor
#### K Neighbors Regressor
#### Decision Tree Regressor
#### Gradient boosting Regressor



In [24]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
import warnings
warnings.filterwarnings('ignore')


In [25]:
df= pd.read_csv(r'C:\Users\user\Downloads\Scaled_Primary.csv_')

In [26]:
df.head()

Year  Completion Rate for Primary Edu  Childhood Education GER  \
0  2012                             99.8                46.379381   
1  2012                             99.7                46.379381   
2  2013                             99.9                46.379381   
3  2013                             99.8                46.379381   
4  2014                             99.9                36.650000   

   Gross enrolment ratio, early childhood educational development programmes,  \
0                                           6.361501                            
1                                           6.361501                            
2                                           6.361501                            
3                                           6.361501                            
4                                           0.000000                            

   Gross intake ratio to the last grade of primary education,  \
0                                             108.19            
1                                             106.88            
2                                             109.76            
3                                             108.90            
4                                             115.17            

   Literacy rate for 25-64 years old  \
0                            73.7825   
1                            73.7825   
2                            73.7825   
3                            73.7825   
4                            73.7825   

   Expenditure on education as a percentage of total government expenditure (%)  \
0                                              19.25                              
1                                              19.25                              
2                                              17.96                              
3                                              17.96                              
4                                              17.32                              

   Government expenditure on education as a percentage of GDP (%)  \
0                                               3.90                
1                                               3.90                
2                                               3.44                
3                                               3.44                
4                                               3.42                

   Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
0           1.0                        0.0                             0.0   
1           1.0                        0.0                             0.0   
2           1.0                        0.0                             0.0   
3           1.0                        0.0                             0.0   
4           1.0                        0.0                             0.0   

   Europe and Northern America  Latin America and the Caribbean  \
0                          0.0                              0.0   
1                          0.0                              0.0   
2                          0.0                              0.0   
3                          0.0                              0.0   
4                          0.0                              0.0   

   Northern Africa and Western Asia  Oceania  Gender_numerical  
0                               0.0      0.0                 1  
1                               0.0      0.0                 0  
2                               0.0      0.0                 1  
3                               0.0      0.0                 0  
4                               0.0      0.0                 1

## Data Transformation

#### 1. Data Rescaling

In [27]:
# Data Rescaling

data_col = df['Year']
df.set_index('Year',inplace =True)

df.head()


Completion Rate for Primary Edu  Childhood Education GER  \
Year                                                             
2012                             99.8                46.379381   
2012                             99.7                46.379381   
2013                             99.9                46.379381   
2013                             99.8                46.379381   
2014                             99.9                36.650000   

      Gross enrolment ratio, early childhood educational development programmes,  \
Year                                                                               
2012                                           6.361501                            
2012                                           6.361501                            
2013                                           6.361501                            
2013                                           6.361501                            
2014                                           0.000000                            

      Gross intake ratio to the last grade of primary education,  \
Year                                                               
2012                                             108.19            
2012                                             106.88            
2013                                             109.76            
2013                                             108.90            
2014                                             115.17            

      Literacy rate for 25-64 years old  \
Year                                      
2012                            73.7825   
2012                            73.7825   
2013                            73.7825   
2013                            73.7825   
2014                            73.7825   

      Expenditure on education as a percentage of total government expenditure (%)  \
Year                                                                                 
2012                                              19.25                              
2012                                              19.25                              
2013                                              17.96                              
2013                                              17.96                              
2014                                              17.32                              

      Government expenditure on education as a percentage of GDP (%)  \
Year                                                                   
2012                                               3.90                
2012                                               3.90                
2013                                               3.44                
2013                                               3.44                
2014                                               3.42                

      Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
Year                                                                            
2012           1.0                        0.0                             0.0   
2012           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2014           1.0                        0.0                             0.0   

      Europe and Northern America  Latin America and the Caribbean  \
Year                                                                 
2012                          0.0                              0.0   
2012                          0.0                              0.0   
2013                          0.0                              0.0   
2013                          0.0                              0.0   
2014                          0.0                              0.0   

      No

In [28]:
scaler = MinMaxScaler()
normalized_df = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)
normalized_df['Year'] = data_col
normalized_df.set_index('Year', inplace =True)
normalized_df.head()

Completion Rate for Primary Edu  Childhood Education GER  \
Year                                                             
2012                         0.997413                 0.285122   
2012                         0.996120                 0.285122   
2013                         0.998707                 0.285122   
2013                         0.997413                 0.285122   
2014                         0.998707                 0.224366   

      Gross enrolment ratio, early childhood educational development programmes,  \
Year                                                                               
2012                                           0.068008                            
2012                                           0.068008                            
2013                                           0.068008                            
2013                                           0.068008                            
2014                                           0.000000                            

      Gross intake ratio to the last grade of primary education,  \
Year                                                               
2012                                           0.845059            
2012                                           0.831545            
2013                                           0.861254            
2013                                           0.852383            
2014                                           0.917062            

      Literacy rate for 25-64 years old  \
Year                                      
2012                           0.713031   
2012                           0.713031   
2013                           0.713031   
2013                           0.713031   
2014                           0.713031   

      Expenditure on education as a percentage of total government expenditure (%)  \
Year                                                                                 
2012                                           0.482729                              
2012                                           0.482729                              
2013                                           0.446206                              
2013                                           0.446206                              
2014                                           0.428086                              

      Government expenditure on education as a percentage of GDP (%)  \
Year                                                                   
2012                                           0.286765                
2012                                           0.286765                
2013                                           0.252941                
2013                                           0.252941                
2014                                           0.251471                

      Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
Year                                                                            
2012           1.0                        0.0                             0.0   
2012           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2014           1.0                        0.0                             0.0   

      Europe and Northern America  Latin America and the Caribbean  \
Year                                                                 
2012                          0.0                              0.0   
2012                          0.0                              0.0   
2013                          0.0                              0.0   
2013                          0.0                              0.0   
2014                          0.0                              0.0   

      No

In [29]:
normalized_df.corr()

Completion Rate for Primary Edu  \
Completion Rate for Primary Edu                                            1.000000   
Childhood Education GER                                                    0.351886   
Gross enrolment ratio, early childhood educatio...                         0.305259   
Gross intake ratio to the last grade of primary...                         0.671671   
Literacy rate for 25-64 years old                                          0.398736   
Expenditure on education as a percentage of tot...                         0.006356   
Government expenditure on education as a percen...                         0.284199   
Central Asia                                                               0.056772   
Central and Southern Asia                                                  0.108787   
Eastern and South-Eastern Asia                                             0.316906   
Europe and Northern America                                                0.267968   
Latin America and the Caribbean                                            0.129879   
Northern Africa and Western Asia                                           0.126447   
Oceania                                                                   -0.691665   
Gender_numerical                                                           0.055559   

                                                    Childhood Education GER  \
Completion Rate for Primary Edu                                    0.351886   
Childhood Education GER                                            1.000000   
Gross enrolment ratio, early childhood educatio...                 0.334212   
Gross intake ratio to the last grade of primary...                 0.377295   
Literacy rate for 25-64 years old                                  0.143463   
Expenditure on education as a percentage of tot...                -0.168636   
Government expenditure on education as a percen...                -0.001142   
Central Asia                                                       0.101699   
Central and Southern Asia                                          0.177365   
Eastern and South-Eastern Asia                                     0.208522   
Europe and Northern America                                        0.043067   
Latin America and the Caribbean                                   -0.049118   
Northern Africa and Western Asia                                   0.063413   
Oceania                                                           -0.355638   
Gender_numerical                                                   0.004159   

                                                    Gross enrolment ratio, early childhood educational development programmes,  \
Completion Rate for Primary Edu                                                              0.305259                            
Childhood Education GER                                                                      0.334212                            
Gross enrolment ratio, early childhood educatio...                                           1.000000                            
Gross intake ratio to the last grade of primary...                                           0.200007                            
Literacy rate for 25-64 years old                                                            0.123545                            
Expenditure on education as a percentage of tot...                                          -0.011862                            
Government expenditure on education as a percen...                                          -0.025302                            
Central Asia                                                                                -0.104493                            
Central and Southern Asia                                                                    0.085812                            
Eastern and South-Eastern Asia                                                               0

In [30]:
# spliting the dataset into train set, test set, validation set.
datax = normalized_df.drop('Completion Rate for Primary Edu', axis = 1)
datay = normalized_df['Completion Rate for Primary Edu']

train_ratio = 0.75
val_ratio = 0.15
test_ratio = 0.10

# making the train set to be 75% of the entire dataset.
x_train, x_test, y_train, y_test = train_test_split(datax, datay, test_size = 1-train_ratio)

# test is now 10% of the entire dataset
# validation is now 15% of the entire dataset

x_val, x_test, y_val, y_test = train_test_split(x_test, y_test, test_size = test_ratio/(test_ratio + val_ratio))

print(x_train, x_val, x_test)

      Childhood Education GER  \
Year                            
2017                 0.404084   
2015                 0.285122   
2018                 0.283002   
2015                 0.568003   
2017                 0.250156   
...                       ...   
2015                 0.582365   
2020                 0.178094   
2021                 0.285122   
2020                 0.447733   
2016                 0.285122   

      Gross enrolment ratio, early childhood educational development programmes,  \
Year                                                                               
2017                                           0.408916                            
2015                                           0.068008                            
2018                                           0.392666                            
2015                                           0.644751                            
2017                                           0.057943           

## Model Training, Testing and evaluation

 
 Because the target variable been 'Completion rate for primary school Edu' is a continoues data and not categorical. therfore, the model algorithm to be use is the regression and not classifiers.

#### 1. Linear Regression


In [31]:

import time

# Record start time
start_time = time.time()

# Create a Linear Regression model
linear_model = LinearRegression()

# Fit the model to the training dataset
linear_model.fit(x_train, y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))





Training took 0.06 seconds.


In [32]:
# Obtain predictions
predicted_values = linear_model.predict(x_test)

#Measuring regression performance
MAE = mean_absolute_error(y_test, predicted_values)
MSE = mean_squared_error(y_test, predicted_values)
RMSE =np.sqrt(mean_squared_error(y_test, predicted_values))
RSS =np.sum(np.square(y_test, predicted_values))
r2=r2_score(y_test, predicted_values)




In [33]:
# Model evaluation

results = pd.DataFrame([['Linear regression', MAE, MSE, RMSE, RSS,r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results

Model       MAE       MSE      RMSE         RSS  r2_score
0  Linear regression  0.121022  0.027553  0.165992  104.785737  0.768865

In [34]:
lreg_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":predicted_values})
lreg_df


actual_values  predictions
Year                            
2019       0.985773     0.971749
2015       0.989653     0.979414
2015       0.979177     0.958788
2017       0.974133     0.948936
2012       1.000000     1.000000
...             ...          ...
2020       0.662700     0.439172
2018       0.383601     0.147149
2012       0.578376     0.334518
2021       0.949172     0.900928
2012       0.948267     0.899210

[170 rows x 2 columns]

#### Decition Tree Regressor

In [35]:
from sklearn.tree import DecisionTreeRegressor

import time


# Record start time
start_time = time.time()

DecisionTree_model = DecisionTreeRegressor()
DecisionTree_model.fit(x_train,y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))


Training took 0.13 seconds.


In [36]:
y_preds = DecisionTree_model.predict(x_test)

In [37]:
# Model Evaluation
mae= mean_absolute_error(y_test, y_preds)
mse= mean_squared_error(y_test, y_preds)
rmse=np.sqrt(mean_squared_error(y_test, y_preds))
rss=np.sum(np.square(y_test, y_preds))
r_squared=r2_score(y_test, y_preds)

In [38]:
model_results = pd.DataFrame([['Decision Tree regression', mae, mse, rmse, rss,r_squared]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)

In [39]:
model_results

Model       MAE       MSE     RMSE         RSS  r2_score
0  Decision Tree regression  0.065356  0.018835  0.13724  104.785737  0.768865

In [40]:
Dreg_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":predicted_values})
Dreg_df


actual_values  predictions
Year                            
2019       0.985773     0.971749
2015       0.989653     0.979414
2015       0.979177     0.958788
2017       0.974133     0.948936
2012       1.000000     1.000000
...             ...          ...
2020       0.662700     0.439172
2018       0.383601     0.147149
2012       0.578376     0.334518
2021       0.949172     0.900928
2012       0.948267     0.899210

[170 rows x 2 columns]

#### K-Nearest Neighbors Regression (KNN Regression)

In [41]:

# Record start time
start_time = time.time()

knn = KNeighborsRegressor()

knn.fit(x_train, y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))




Training took 0.16 seconds.


In [43]:
knn_predictions  = knn.predict(x_test)

In [44]:
print(knn_predictions.shape, y_test.shape)

(170,) (170,)


In [48]:
# getting model scores
knn_r2 = r2_score(y_test, knn_predictions)
knn_mse = mean_squared_error( y_test,knn_predictions)
knn_rmse = np.sqrt(knn_mse)
knn_mae = mean_absolute_error(y_test,knn_predictions)
knn_rss = np.sum(np.square(y_test, knn_predictions))

In [49]:
model_results = pd.DataFrame([['KNN regression', knn_mae, knn_mse, knn_rmse, knn_rss,knn_r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)

In [50]:
model_results

Model       MAE       MSE      RMSE         RSS  r2_score
0  KNN regression  0.082342  0.018448  0.135824  104.785737  0.793223

In [51]:
knn_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":knn_predictions})
knn_df.sort_index(ascending=True, inplace=True)
knn_df

actual_values  predictions
Year                            
2012       0.948267     0.899210
2012       0.983187     0.966656
2012       0.624935     0.390544
2012       0.578376     0.334518
2012       0.204604     0.041863
...             ...          ...
2021       0.974521     0.949692
2021       0.967926     0.936880
2021       0.949172     0.900928
2021       1.000000     1.000000
2021       0.945810     0.894556

[170 rows x 2 columns]

In [52]:
from sklearn.ensemble import GradientBoostingRegressor


# Record start time
start_time = time.time()

Gbr = GradientBoostingRegressor()
Gbr.fit(x_train,y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))



Training took 0.68 seconds.


In [54]:
Gbr_predictions  = Gbr.predict(x_test)

In [55]:
# getting model scores
Gbr_r2 = r2_score(y_test, Gbr_predictions)
Gbr_mse = mean_squared_error( y_test,Gbr_predictions)
Gbr_rmse = np.sqrt(Gbr_mse)
Gbr_mae = mean_absolute_error(y_test,Gbr_predictions)
Gbr_rss = np.sum(np.square(y_test, Gbr_predictions))

In [56]:
model_results = pd.DataFrame([['GBR regression', knn_mae, knn_mse, knn_rmse, knn_rss,knn_r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)

In [57]:
model_results

Model       MAE       MSE      RMSE         RSS  r2_score
0  GBR regression  0.082342  0.018448  0.135824  104.785737  0.793223

In [58]:
Gbr_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":knn_predictions})
Gbr_df.sort_index(ascending=True, inplace=True)
Gbr_df

actual_values  predictions
Year                            
2012       0.948267     0.899210
2012       0.983187     0.966656
2012       0.624935     0.390544
2012       0.578376     0.334518
2012       0.204604     0.041863
...             ...          ...
2021       0.974521     0.949692
2021       0.967926     0.936880
2021       0.949172     0.900928
2021       1.000000     1.000000
2021       0.945810     0.894556

[170 rows x 2 columns]

In [59]:
# Printing the stored results of all the three models
results.drop_duplicates()


Model       MAE       MSE      RMSE         RSS  \
0         Linear regression  0.121022  0.027553  0.165992  104.785737   
1  Decision Tree regression  0.065356  0.018835  0.137240  104.785737   
2            KNN regression  0.082342  0.018448  0.135824  104.785737   
3            GBR regression  0.082342  0.018448  0.135824  104.785737   

   r2_score  
0  0.768865  
1  0.768865  
2  0.793223  
3  0.793223

### conclusion

1.Linear Regression time for the model to train is 0.06 seconds

1.Decision Tree Regression seems to perform well in terms of minimizing errors and time for the model to tarin is 0.13 seconds.

2.KNN Regression and GBR Regression have higher R-squared values, suggesting better explanatory power.
also, the time for the two models to train are 0.16 and 0.68 respectively.

3.Both KNN and GBR have the same MAE, MSE, and RMSE, indicating similar performance on average.